File: mergeSort.js
// PROBLEM: sort an array
// aproach: meerging sorted array is easy
// recurse down to single element arrays -- which are sorted arrays
// then buuld back up, merging sorted arrays

function mergeSort(arr) {
  
  if (arr.length === 1) {
    return arr;
  }

  // split array to left and right
  let mid = Math.floor(arr.length / 2)
  let leftArr = arr.slice(0, mid);
  let rightArr = arr.slice(mid);

  // recursive call on left and right
  leftArr = mergeSort(leftArr)
  rightArr = mergeSort(rightArr);

  // when they return, they are sorted
  // now merge
  return merge(leftArr, rightArr)
}

// helper function
function merge(left, right) {
  let sorted = [];

  // exit consition is empty array b/c we are mutating array
  // mutating the arry is fine b/c we created left and right as new arrays
  // we are not mutating the argument passed o mergeSort
  while (left.length && right.length) {
    if (left[0] < right[0]) {
      sorted.push(left.shift());
    } else {
      sorted.push(right.shift());
    }
  }
  // return an array + empty array + remainder (or vice versa)
  return [...sorted, ...left, ...right]
}


console.log(mergeSort([1,7,6,3,2,9]));

File: mergeSort.py



# arr = [2, 4, 7, 8, 1, 3, 9]    
# print(merge_sorted_arrays(arr, 0, 3, 6))

def sort(nums):
    def merge_sorted_arrays(nums, l, m, r):
        """

        O: same array
        """

        if l == r:
            return nums
        i = l
        j, k = 0, 0

        left = nums[l : m + 1]
        right = nums[m + 1 : r + 1]

        while j < len(left) and k < len(right):
            if left[j] <= right[k]:
                nums[i] = left[j]
                j += 1
            else:
                nums[i] = right[k]
                k += 1
            i += 1

        while j < len(left):
            nums[i] = left[j]
            j += 1
            i += 1

        while k < len(right):
            nums[i] = right[k]
            k += 1
            i += 1

        return nums    
    
    def merge_sort(nums, l, r):
        if (l == r):
            return nums

        m = (l + r) // 2
        merge_sort(nums, l, m)
        merge_sort(nums, m + 1, r)
        
        return merge_sorted_arrays(nums, l, m, r)

    return merge_sort(nums, 0, len(nums) - 1)



arr = [2, 10, 4, 0, 7, 8, 1, 3, 9]    
print(sort(arr))